submitting-changes.chapter.md: explain *why* we have three branches (#225920)
NixOS/nixpkgs: doc/contributing/submitting-changes.chapter.md

- Hydra builds for master and staging should not be used as testing platform, it’s a build farm for changes that have been already tested.
- When changing the bootloader installation process, extra care must be taken. Grub installations cannot be rolled back, hence changes may break people’s installations forever. For any non-trivial change to the bootloader please file a PR asking for review, especially from \@edolstra.

### Branches

The nixpkgs repo has three major branches: `master`, `staging`, and
`staging-next`.

The most important distinction between them is that `staging`

> (colored red in the diagram below) can receive commits which cause
> a mass-rebuild (for example, anything that changes the `drvPath` of
> `stdenv`).  The other two branches `staging-next` and `master`
> (colored green in the diagram below) can *not* receive commits which

Reviewer: Colors are transient and not even visible to everyone, depending on output medium or anatomy. Also another place for double book keeping to keep the diagram and the text in sync.

We should clarify what a mass rebuild is without going into implementation details that aren't relevant here. Tried to incorporate that into the suggestion.

```suggestion
can receive commits that cause
a mass-rebuild, that is, changes to data a large amount of derivations depend on and therefore require those to be rebuilt.
The other two branches, `staging-next` and `master`, must *not* receive commits which
```

Author: Yes, I'm aware that colorblind people exist, which is one of the reasons why this PR switches from *pastels* to primary colors.

Mentioning the fact that certain parts of the diagram are certain colors does not hurt, even if a few people cannot see those colors.

cause a mass-rebuild.

Arcs between the branches show possible merges into these branches,
either from other branches or from independently submitted PRs.  The
colors of these edges likewise show whether or not they could
trigger a mass rebuild (red) or must not trigger a mass rebuild
(green).

Hydra runs automatic builds for the green branches.

Notice that the automatic merges are all green arrows.  This is by
design.  Any merge which might cause a mass rebuild on a branch
which has automatic builds (`staging-next`, `master`) will be a
manual merge to make sure it is good use of compute power.

Nixpkgs has two branches so that there is one branch (`staging`)
which accepts mass-rebuilding commits, and one fast-rebuilding
branch which accepts independent PRs (`master`).  The `staging-next`

> branch allows the Hydra operators to "batch up" groups of commits to

Reviewer: ```suggestion
branch allows the Hydra operators to batch groups of commits to
```

`staging` to be built.  By keeping the `staging-next` branch

> separate from `staging`, this "batching" decision does not block

Reviewer: ```suggestion
separate from `staging`, this batching does not block
```

Nothing peculiar about batching that seems to warrant putting it in quotes.
